logserver/src/main/java/alluxio/logserver/AlluxioLogServerProcess.java

```diff
@@ -0,0 +1,228 @@
+  }
+
+  @Override
+  public void waitForReady() {
+    CommonUtils.waitFor(this + " to start", new Function<Void, Boolean>() {
+      @Override
+      public Boolean apply(Void input) {
+        return mStopped == false;
+      }
+    }, WaitForOptions.defaults().setTimeoutMs(10000));
+  }
+
+  /**
+   * Create and start logging server and client thread pool.
+   */
+  private void startServing() {
+    SynchronousQueue<Runnable> synchronousQueue =
+        new SynchronousQueue<>();
+    mThreadPool =
+        new ThreadPoolExecutor(mMinNumberOfThreads, mMaxNumberOfThreads,
+            STOP_TIMEOUT_IN_MS, TimeUnit.MILLISECONDS, synchronousQueue);
+    try {
+      mServerSocket = new ServerSocket(mPort);
+    } catch (IOException e) {
+      LOG.error("Failed to bind to port {}.", mPort);
+      return;
+    }
+    mStopped = false;
+    while (!mStopped) {
+      try {
+        Socket client = mServerSocket.accept();
+        InetAddress inetAddress = client.getInetAddress();
+        AlluxioLog4jSocketNode clientSocketNode =
+            new AlluxioLog4jSocketNode(this, client);
+        RetryPolicy retryPolicy = new ExponentialBackoffRetry(
+            BASE_SLEEP_TIME_IN_MS, MAX_SLEEP_TIME_IN_MS, MAX_NUM_RETRY);
+        while (true) {
+          try {
+            mThreadPool.execute(clientSocketNode);
+            break;
+          } catch (Throwable t) {
+            if (t instanceof RejectedExecutionException) {
+              if (!retryPolicy.attemptRetry()) {
+                LOG.warn("Connection with {} has been rejected by ExecutorService {} times"
+                    + "till timedout, reason: {}",
+                    inetAddress.getHostAddress(), retryPolicy.getRetryCount(), t);
+                client.close();
+                break;
+              }
+            } else if (t instanceof Error) {
+              LOG.error("ExecutorService threw error: ", t);
+              throw (Error) t;
+            } else {
+              LOG.warn("ExecutorService threw error: ", t);
+              break;
+            }
+          }
+        }
+      } catch (IOException e) {
+        if (!mStopped) {
+          LOG.warn("Socket transport error occurred during accepting message.", e);
+        }
+        break;
+      }
+    }
+  }
+
+  /**
+   * Called from another thread (actually in the shutdown hook of {@link AlluxioLogServer}
+   * to stop the main thread of {@link AlluxioLogServerProcess}.
+   */
+  private void stopServing() {
+    mStopped = true;
+    if (mServerSocket != null) {
+      try {
+        mServerSocket.close();
+      } catch (IOException e) {
+        LOG.warn("Exception in closing server socket.", e);
+      }
+    }
+
+    mThreadPool.shutdown();
+    long timeoutMS = STOP_TIMEOUT_IN_MS;
+    long now = System.currentTimeMillis();
+    while (timeoutMS >= 0) {
+      try {
+        mThreadPool.awaitTermination(timeoutMS, TimeUnit.MILLISECONDS);
+        break;
+      } catch (InterruptedException e) {
```

Comment:
`InterruptedException` means that another thread has tried to interrupt this one by setting the thread's interrupt flag. When the exception is generated, the flag is unset because it's assumed that the catcher of `InterruptedException` will do the right thing to terminate the thread. Whenever you handle `InterruptedException`, you need to either reset the interrupt flag (`Thread.currentThread().interrupt()`), re-throw `InterruptedException`, or cause the thread to exit. In this situation I suggest handling via

```java
  Thread.currentThread.interrupt();
  throw new RuntimeException(e);
```

Reply:
Done.